torch/export/_unlift.py

```diff
@@ -1,273 +1,301 @@
-            # In the case that the same node is returned multiple times,
-            # node.all_input_nodes will only iterate that node once
-            for return_node in pytree.tree_flatten(node.args)[0]:
-                return_node_name = return_node.name
-                # we found a param/buffer mutation
-                if return_node_name in buffers_to_mutate:
-                    # TODO Fix situation here to replace dot with underscore...
-                    buffer_node_name = buffers_to_mutate[return_node_name].replace(
-                        ".", "_"
-                    )
-                    assert buffer_node_name in buffer_name_to_node
-                    buffer_node = buffer_name_to_node[buffer_node_name]
-                    with gm.graph.inserting_before(node):
-                        buffer_update_node = gm.graph.call_function(
-                            torch.ops.aten.copy_.default, (buffer_node, return_node)
-                        )
-                else:
-                    user_output_nodes.append(return_node)
-            with gm.graph.inserting_before(node):
-                # Only return user outputs
-                new_output = gm.graph.output(tuple(user_output_nodes))
-                node.replace_all_uses_with(new_output)
-                gm.graph.erase_node(node)
-
-    # Step 3: Fix the input/output of the graph now that we deleted
-    # some args.
-    gm.graph.lint()
+def _unlift_inputs_as_getattr(
+    gm: torch.fx.GraphModule,
+    lifted_inputs: List[Optional[str]],
+) -> Tuple[Dict[str, torch.fx.Node], Dict[str, torch.fx.Node]]:
+    """
+    Unlift inputs referring to params/buffers/constants as getattr nodes in the
+    graph
+    """
+    unlifted_name_to_node = {}
+    input_name_to_node = {}
+
+    placeholder_nodes = [node for node in gm.graph.nodes if node.op == "placeholder"]
+    assert len(lifted_inputs) == len(placeholder_nodes)
+    for input_node, lifted_node in zip(placeholder_nodes, lifted_inputs):
+        if lifted_node is None:
+            input_name_to_node[input_node.name] = input_node
+
+        else:
+            with gm.graph.inserting_after(input_node):
+                getattr_node = gm.graph.get_attr(lifted_node.replace(".", "_"))
+                input_node.replace_all_uses_with(getattr_node)
+                metadata = input_node.meta
+                gm.graph.erase_node(input_node)
+                getattr_node.meta = metadata
+                unlifted_name_to_node[lifted_node.replace(".", "_")] = getattr_node
+
+    return unlifted_name_to_node, input_name_to_node
+
+
+def _insert_copy_for_mutations(
+    gm: torch.fx.GraphModule,
+    mutated_outputs: List[Optional[str]],
+    unlifted_name_to_node: Dict[str, torch.fx.Node],
+    input_name_to_node: Dict[str, torch.fx.Node],
+) -> None:
+    """
+    Find the all the buffers and inputs that were mutated and insert copy_
+    operators to reflect mutations.
+    """
+    output_node = list(gm.graph.nodes)[-1]
```

Comment:
I don't think this is guaranteed; while graphs generated by make_fx/dynamo almost certainly have this property, I remember someone tried to assert this in fbcode and it was tripped. We should probably just search for output to be safe.